src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala

```diff
@@ -254,7 +258,9 @@ class MarathonSchedulerService @Inject() (
           electionService.abdicateLeadership(error = result.isFailure, reoffer = isRunningLatch.getCount > 0)
         }
 
-        driver = None
+        synchronized {
+          driver = None
+        }
 
         log.info(s"Call postDriverRuns callbacks on ${prePostDriverCallbacks.mkString(", ")}")
         Await.result(Future.sequence(prePostDriverCallbacks.map(_.postDriverTerminates)), config.zkTimeoutDuration)
```

Comment:
Hey @timcharper, 
I was out for holidays and didn't get a chance to checkout these changes.  I added comments in your code ***
``` diff
+        val driverCleared = synchronized {
+          val cleared = driver.nonEmpty
+          driver = None
+          cleared
+        }
+

// *** if we get here and driver was not cleared in previous check, but leader latch has now started it's shutdown we 
// ***  may still get a deadlock in abdicateLeadership

         // ONLY do this if there's some sort of driver crash: avoid invoking abdication logic if
         // the driver was stopped via stopDriver. stopDriver only happens when
         //   1. we're being terminated (and have already abdicated)
         //   2. we've lost leadership (no need to abdicate if we've already lost)
-        driver.foreach { _ =>
+        if (driverCleared) {
           // tell leader election that we step back, but want to be re-elected if isRunning is true.
           electionService.abdicateLeadership(error = result.isFailure, reoffer = isRunningLatch.getCount > 0)
         }
```

I think your changes will shrink the window where we could get a deadlock, but I think there is still a slight chance.  That is why we overrode the leadership functions in the CuratorElectionService, so the locks will block each other out completely.